File: 0131-palindrome-partitioning/0131-palindrome-partitioning.cpp
class Solution {
 
public:

   bool check(string str){
    string str2 = str;
    reverse(str2.begin(),str2.end());
    if(str == str2){
        return true;
    }
    return false;
    }
    void help(vector<vector<string>>& ans, vector<string>& temp,string str){
       if(str.size() == 0){
        ans.push_back(temp);
        return;
       }

      for(int i=0; i<str.size();i++){
         
         string s = str.substr(0,i+1);
         if(check(s)){
           temp.push_back(s);
           help(ans,temp,str.substr(i+1));
           temp.pop_back();
         }
      }


    }

    vector<vector<string>> partition(string s) {
        vector<vector<string>>ans; vector<string> temp;
        help(ans,temp,s);
        return ans;
    }
};

File: 0131-palindrome-partitioning/0131-palindrome-partitioning.java
class Solution {
    boolean check(String s){
        int st= 0, end = s.length()-1;

        while(st<end){
            if(s.charAt(st) != s.charAt(end)) return false;
            st++;
            end--;
        }

        return true;
    }
    void help(int idx ,String s, List<List<String>> ans, List<String> temp){
        if(idx == s.length()){
            ans.add(new ArrayList<>(temp));
            return;
        }
        for(int i=idx ;i<s.length();i++){
          if(check(s.substring(idx,i+1))){
            temp.add(s.substring(idx,i+1));
            help(i+1,s,ans,temp);
            temp.remove(temp.size()-1);
          }
         
        }
       return;
    }
    public List<List<String>> partition(String s) {
         List<List<String>> ans = new ArrayList<>();
       List<String> temp = new ArrayList<>();
       help(0,s,ans,temp);
       return ans;
    }
}